Add parser for mensaar.de api

ialokim opened the pull request:
This PR adds support for the API provided by [mensaar.de](https://mensaar.de/) and thus for the following universities:
* Mensa / Mensacafe Saarbrücken
* Mensa Homburg
* Cafeteria HTW Göttelborn

Fixes #1

ialokim commented:
Up to now, only support the German language version of the API, I'm not really sure how to expose a translated menu to OpenMensa :thinking:

fefrei commented:
In practice, most strings aren't actually translated anyway (yet) because of internal difficulties `¯\_(ツ)_/¯`

fefrei commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+    'm': 'employee'
+}
+
+SPECIAL_NOTICES = [
```
I don't really like that you need to hard-code certain notice keys here. Maybe we can extend the `getBaseData` call to provide you with the categorization we need? What would we need to add, exactly?

ialokim replied:
I'm building strings to use as notes on the meal object (see lines [58](https://github.com/mswart/openmensa-parsers/pull/74/files/8a2c2afd2ba9c57f951bc042f568b9529dc39cda#diff-91dab909c61ef0b60c3823e035632147R58) and [68](https://github.com/mswart/openmensa-parsers/pull/74/files/8a2c2afd2ba9c57f951bc042f568b9529dc39cda#diff-91dab909c61ef0b60c3823e035632147R68)), because of a pretty limited feed specification (see [openmensa#34](/openmensa/openmensa/issues/34)) with no differentiation between allergic, nutritional or ingredients' information.

Regarding the mensaar.de API, the only differentiation between `notices` is provided by the `isAllergen` value, but most of the `notices` are nouns (and could be formatted as in the lines linked to above), but these "special" ones are adjectives that does not fit within the string. I therefore needed to handle them separately to avoid a `note` like `mit geschwefelt`, for example.

ialokim replied:
I'm not really sure whether it is necessary to extend the API, but it would be a nice improvement to mark these cases differently.

fefrei replied:
Hm, in that case, I think your list is incomplete: “mit kann Spuren von Nüssen beinhalten” isn't right either, but `nu` is not in your exception list.

I'd be happy to extend the API here, but I currently don't see which exact criterion I'd have to set up to help you and others with this problem. Given that English translations aren't currently available anyways, maybe checking whether the notice text starts with a capital letter might be enough? The only exception to this rule seems to be `Aus zum Teil fein zerkleinertem Fleisch`, and I'm removing the capital letter right now 😉

ialokim replied:
> Hm, in that case, I think your list is incomplete: “mit kann Spuren von Nüssen beinhalten” isn't right either, but nu is not in your exception list.

Uuuuh, you're right, I've missed that one! Seems like there are no meals with nuts within the next four days :laughing: 

> Given that English translations aren't currently available anyways, maybe checking whether the notice text starts with a capital letter might be enough?

That would be an alternative, indeed. I'm not sure which one I prefer, though.

> I'd be happy to extend the API here, but I currently don't see which exact criterion I'd have to set up to help you and others with this problem.

Just thinking, if you would introduce another key called `isAdjective` or `isDescriptive` or something similar... or a key for the nouns called `isComponent` or `isIngredient`...?

fefrei replied:
I'm totally open to that, but I haven't found something that sounds totally consistent yet. For example, *kann Spuren von Nüssen beinhalten* is in some sense an ingredient, but for your purpose, you'd clearly want `isIngredient: false`, right?

There are even more corner-cases: What about *Krebstiere, Krusten- und Schalentiere*? These are clearly ingredients, or components, or however you want to call them, but you cannot say *mit Krebstiere, Krusten- und Schalentiere* – you'd need to say *mit Krebstiere**n**, Krusten- und Schalentiere**n***.

Another workaround would be for the API to always return an alternative “descriptive form”, which would never be a single substantive, but always a describing phrase. But that also doesn't feel like a good solution for me.

ialokim replied:
>  For example, kann Spuren von Nüssen beinhalten is in some sense an ingredient, but for your purpose, you'd clearly want isIngredient: false, right?

Is that a possible future example? I didn't get it by the API up to now :thinking: 
I think these kind of cases shouldn't have a `displayName` as _kann Spuren von Nüssen beinhalten_ but only _Nüsse_ and another key, some sort of `isPossibility` or `couldBe`... It would be up to the client to handle this case accordingly.

> What about Krebstiere, Krusten- und Schalentiere?

In this case, I've been lucky because it's marked as `isAllergen` and _enthält Krebstiere, Krusten- und Schalentiere_ works. But that's the problem of the German language and Its accusative forms :see_no_evil: 

> Another workaround would be for the API to always return an alternative “descriptive form”, which would never be a single substantive, but always a describing phrase. But that also doesn't feel like a good solution for me.

Nor to me, to be honest. The best possible solution IMHO would be to restrict the possible values of `displayName` (e.g. to one word) and to abstract other properties to standard keys. But that implies some changes on client-side (e.g. _ohne Schweinefleisch_ should only have _Schweinefleisch_ as `displayName`, while the client has to recognize the `isNegative` key and add an _ohne_ to the displayed form).

Or to keep backward compatibility, you could introduce a new key `name` with the restriction mentioned above and keep the `displayName` key as a sort of "descriptive form" (as it already is in some cases).

fefrei replied:
> Is that a possible future example? I didn't get it by the API up to now 🤔

The `isIngredient` flag was just an example as I'm not sure what the best option is (yet).

> I think these kind of cases shouldn't have a displayName as kann Spuren von Nüssen beinhalten but only Nüsse and another key, some sort of isPossibility or couldBe... It would be up to the client to handle this case accordingly.

Hm, there is some truth to that, but that would be a breaking change (at least semantically), and this text was provided like so from Studentenwerk, so I don't really want to change it.

> In this case, I've been lucky because it's marked as isAllergen and enthält Krebstiere, Krusten- und Schalentiere works. But that's the problem of the German language and Its accusative forms 🙈

You're a lucky guy 😉

>  abstract other properties to standard keys

That sounds good, but it's actually a bit tricky. There are at least two cases (*components* and *properties*, although the separation is sometimes a bit tricky), negation, *and* and *or, ...

ialokim replied:
> The isIngredient flag was just an example as I'm not sure what the best option is (yet).

I was referring to the concrete example of _kann Spuren von Nüssen enthalten_ :smile: 

> Hm, there is some truth to that, but that would be a breaking change (at least semantically), and this text was provided like so from Studentenwerk, so I don't really want to change it.

That's why I thought of leaving this text provided from Studentenwerk as it is for `displayName` and introducing the new (restricted) `name` key.

> There are at least two cases (components and properties, although the separation is sometimes a bit tricky)

In which cases would it be tricky?

> and and *or, ...

I'm not sure what do you mean by _and_ and _or_?

fefrei replied:
> I was referring to the concrete example of kann Spuren von Nüssen enthalten 😄

Oops. Turns out the configuration file provided by Studentenwerk contains both

    "nu": {
        "displayName": {"de": "Schalenfrüchte (Nüsse)", "en": "Nuts"},
        "isAllergen": true
    },

and

    "nu": {
        "displayName": {"de": "kann Spuren von Nüssen beinhalten", "en": "may contain traces of nuts"},
        "isAllergen": false
    },

so we actually only see the former.

> In which cases would it be tricky?

How would you classify *aus zum Teil fein zerkleinertem Fleisch*? In some sense, it's both component and property.

> I'm not sure what do you mean by and and or?

I'm talking about weird notices like *Macadamia-oder-Queenslandnüsse* or *Krebstiere, Krusten- und Schalentiere*. None of these are made up, unfortunately 😟

ialokim replied:
> How would you classify aus zum Teil fein zerkleinertem Fleisch? In some sense, it's both component and property.

Okay, I see. So what do you think about introducing the three different notices groups _components_, _properties_ and _other_, while _other_ would cover these (very few) special cases. So, every _notice_ object would have a key `group` or (`type`) with the only accepted values `component` (or `ingredient`), `property` and `other` (or simply `note`).

> I'm talking about weird notices like Macadamia-oder-Queenslandnüsse or Krebstiere, Krusten- und Schalentiere.

Hmmmm, then why not better leaving these notices' names together? If they are only composed of nouns and _und_/_oder_ :thinking:

fefrei replied:
Hm, that could work, although `other` still feels like cheating. I've asked the IT department of Studentenwerk for their opinion, maybe they have some comments.

ialokim commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+                notices_allergens.append(
+                    'enthält %s' % notice['displayName'])
+        else:
+            if notice_short == 'nsf':  # ohne Schweinefleisch
```
Another special case is `nsf`, the only value that states explicitly the _absence_ of an ingredient, so here too, I had to avoid a `note` as `mit ohne Schweinefleisch`.

fefrei replied:
Indeed, I'd say this is an oversight in the API. It's fixed now:

```
        "nu": {
            "displayName": "Schalenfrüchte (Nüsse)",
            "isAllergen": true,
            "isNegated": false
        },
        "nsf": {
            "displayName": "ohne Schweinefleisch",
            "isAllergen": false,
            "isNegated": true
        },
```

I hope this helps!

ialokim replied:
Wow, you're quick! Thanks a lot!

fefrei replied:
That's what happens when the API isn't operated by the Studentenwerk itself, but an enthusiastic student – API changes are just one push away 🎉

klemens replied:
Could you incorporate this change?

ialokim replied:
Done.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def build_notes(notices, components):
+    global base_data
```
`base_data` is not used in this function.

ialokim replied:
You're right, I've removed it.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+        if len(notices_others) > 1:
+            notices_others[-2] = (
+                '%s und %s' % (notices_others[-2], notices_others[-1]))
+            del notices_others[-1]
```
As both these do the same thing, you could mode this into a separate function. Also, I am not sure if separating the last two by „und“ is really necessary, especially as single notes can already contain „und“:
> dazu Chili-Kräuterdip (enthält Soja, Milch und Laktose, Senf und Weizen)

ialokim replied:
Okay, I've changed it to a little sub-function. However, I would like to keep the separating by an "und" for the last item.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+    notices_others_last = []
+
+    for notice_short in notices:
+        notice = base_data['notices'][notice_short]
```
Are all notices always contained in the base data?

fefrei replied:
Yes, every notice must be specified in the `getBaseData` call. (An exception would be if the base data is stale. This is very unlikely, as new notices are introduced rarely.)

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+
+            for meal in counter['meals']:
+                if 'knownMealId' in meal:
+                    print('knownMealId: %s' % meal['knownMealId'])
```
There is currently one known meal in the base data:
```json
"knownMeals": {
  "gebfl": {
    "displayName": "Gebackener Fleischkäse"
  }
}
```

Should this be used somehow? Do these meals maybe have no displayName, which would cause the `counter['displayName']` above to fail?

fefrei replied:
This is meant to allow recognizing recurring meals, for features like marking meals as favorites. (Simply matching the title isn't enough because there often are minor differences.)

So far, I couldn't actually convince Studentenwerk Saarland to build the list of recurring meals, so this list only contains this single entry as an example.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+        print('New unknown price tier: %s (displayName: %s)' %
+              (role, data['priceTiers'][role]))
+        print('All prices for this price tier will be ignored.')
+        print('Please consider updating the parser!')
```
I don't think printing these notes (also above) really helps as probably no one notices these anyway. You could throw an exception, which will get noticed, as openmensa.org sends emails when parsers fail, or simply ignore unknown roles.

ialokim replied:
Okay, I've changed it to throw an error instead.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+
+
+with urlopen(URL_BASE + URL_BASE_DATA) as response:
+    data = json.loads(response.read().decode())
```
Requesting this in the main scope is not a good idea. It leads to executing network requests when simply including the module, which happens even if the parser is not used.
Also depending on how this is hosted, it might be that this is executed once when starting the application and then never updated again.

So I suggest simply moving this into a separate function that gets called in `parse_url` and define the parsers manually (they currently (see #78) have to be defined manually in openmesa.org anyway and they probably don't change that often). This way you can also avoid the `global` definitions.

ialokim replied:
Humm, okay, so you suggest hardcoding the different locations once so that they cannot be updated automatically when the `BaseData` has been updated? Or am I missing some way this would be possible without loading the `BaseData` in the main scope?

klemens replied:
> […] hardcoding the different locations once […]

Yes, I think this is the best solution for now.

> Or am I missing some way this would be possible without loading the BaseData in the main scope?

I don't think this is possible right now, but we can certainly discuss adding a feature that allows fetching the canteens dynamically, maybe together with support for actually creating the new canteens at openmensa.org automatically (through extended metadata).

ialokim replied:
Okay, I've changed it now.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+                meal_notes = (
+                    build_location(counter_description) +
+                    build_hours(counter_hours) +
+                    build_notes(meal['notices'], meal['components']))
```
Just a quick info: While the generated xml keeps the notes in order, openmensa.org doesn't store them in order (see #77 for discussion). This is also something that we might want to change (I haven't really looked into this yet).

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def build_hours(opening_hours):
+    if not opening_hours:
```
NIT: This should be `if opening_hours is None`.

ialokim replied:
Changed.

fefrei commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+URL_BASE_DATA = "getBaseData"
+URL_MENU = "getMenu/"
+
+UTC_DATE_STRING = "%Y-%m-%dT%H:%M:%S.%fZ"
```
Yes. As per the spec, the `date` field on a *day* object contains *a date representing the corresponding day at midnight, UTC*.

Y0hy0h replied:
Maybe you could add a comment linking to the spec?

fefrei replied:
The specification isn't public (so far).

Y0hy0h replied:
Oh, ok. Maybe just a comment stating the reliance on the spec? Just so that when it changes and someone reads the code, there is an explicit dependency stated, so that they'll catch it more likely.

Something like `Mensaar.de always returns an UTC date` or similar? Or what do you think of such a comment?

ialokim replied:
I agree, it's added.

fefrei commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+                    meal_name = '%s: %s' % (meal['category'], meal_name)
+
+                meal_notes = (
+                    build_location(counter_description) +
```
Typically yes, but the API doesn't promise that, this might also contain other human-readable information.

ialokim replied:
Okay, I'm not quite sure how to handle this case of checking whether it is a location or not. The `build_location` function could check for all known locations prior to prepending "Ort: ", but this would be hardcoded information again.

So better drop the `build_location` logic or is there a way changing the API to always get a location with a predefined key?

klemens replied:
I think we can leave it for now, if this is typically the location and change it if there are actually other types of descriptions someday.

ialokim replied:
I've added a little comment explaining this.

ialokim commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+            counter_name = counter['displayName']
+            counter_description = counter['description']
+            counter_hours = counter.get('openingHours')
+            counter_color = counter.get('color')
```
You're right, I didn't find a way to reflect color information in openmensa feeds. I'll remove it.

ialokim commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+        date = day.get('date')
+
+        if today and (datetime.datetime.now().date() !=
+                      datetime.datetime.strptime(date, UTC_DATE_STRING).date()):
```
Done.

ialokim commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+    for day in data['days']:
+        date = day.get('date')
+
+        if today and (datetime.datetime.now().date() !=
```
Thanks, changed it accordingly.

ialokim commented on parsers/saarland.py:
```diff
@@ -0,0 +1,199 @@
+        data = json.loads(response.read().decode())
+
+    for day in data['days']:
+        date = day.get('date')
```
You're right, thanks!

ialokim commented:
I've now addressed most of the issues, thanks a lot for your review!

Please see my comment concerning the loading of the base data inside the main scope.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,223 @@
+    for role in data['priceTiers']:
+        if role not in ROLES:
+            # Found an unknown price tier
+            # All prices for this price tier will be ignored.
```
This line of the comment is no longer accurate.

ialokim replied:
Removed it.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,223 @@
+                    # This is meant to allow recognizing recurring meals,
+                    # for features like marking meals as favorites.
+                    # Up to now, not really used in the mensaar.de API.
+                    print('knownMealId: %s' % meal['knownMealId'])
```
I's fine if you want to keep it, but please print to stderr with `file=sys.stderr`.

ialokim replied:
Okay, I've changed it.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,223 @@
+            base_data['roles'][role] = ROLES[role]
+
+
+base_data = {}
```
Is there is specific reason for storing this data in a global variable? If not I would prefer just returning the data from the function and passing it to the other functions as needed.

ialokim replied:
No, there is no reason. I've changed it.

klemens commented on parsers/saarland.py:
```diff
@@ -0,0 +1,223 @@
+                        if role in base_data['roles']:
+                            meal_prices[base_data['roles'][role]] = prices[role]
+                        else:
+                            print('Ignoring price for unknown role %s.' % role)
```
I guess this cannot happen anymore, as we now abort the parser when there are new price tiers, so we could remove the check and print. If you want to keep the print, please also print to stderr like above.

ialokim replied:
Removed the check.

klemens commented:
Thanks again! I just merged the parser. It should become available under http://omfeeds.devtation.de/saarland/sb/today.xml etc. really soon.

Do you have an openmensa.org account (or want to create one) and add the canteens? If so, you have to add a new "parser" there and add the canteens using the same name as in the parser here. You can then add the index url (http://omfeeds.devtation.de/saarland/index.json) to the "parser" and press update, which will add the right feed urls automatically.

ialokim commented:
> Thanks again! I just merged the parser. It should become available under http://omfeeds.devtation.de/saarland/sb/today.xml etc. really soon.

Thanks a lot!

> Do you have an openmensa.org account (or want to create one) and add the canteens? If so, you have to add a new "parser" there and add the canteens using the same name as in the parser here. You can then add the index url (http://omfeeds.devtation.de/saarland/index.json) to the "parser" and press update, which will add the right feed urls automatically.

I did this, but I think there might be some problem, because it does not show feeds in the parser details. Also, if I click the "update sources with index url" button, it says "5 new sources, 0 sources added". Opening the index url right away in the browser gives an error page. Not sure what I'm doing wrong?

klemens commented:
> Opening the index url right away in the browser gives an error page

This is due to the server returning the wrong mime-type. You can view the page source to work around that problem.

> I did this, but I think there might be some problem, because it does not show feeds in the parser details

Have you manually added the canteens (including address etc)? Currently they cannot be automatically added by the index url, because the metadata returned only contains information about the feeds and not the canteens itself.

ialokim commented:
> Have you manually added the canteens (including address etc)? Currently they cannot be automatically added by the index url, because the metadata returned only contains information about the feeds and not the canteens itself.

Okay, got it. But it seems the omfeeds.devtation.de server is offline now.

Btw, it seems they added two more canteens, I should open a new PR adding these, right?

klemens commented:
> But it seems the omfeeds.devtation.de server is offline now.

It is online again.

> Btw, it seems they added two more canteens, I should open a new PR adding these, right?

Yes, please do so!

ialokim commented:
> It is online again.

Okay, but there still seems to be an error, loading http://omfeeds.devtation.de/saarland/sb/today.xml gives a 502 Bad Gateway page. :thinking:

klemens commented:
It fails because of the new locations:
> RuntimeError: Unknown location: htwcas (displayName: {'displayName': 'Mensa HTW Saar CAS', 'description': '…'})

Maybe we shouldn't fail hard when there are new locations.

ialokim commented:
> It fails because of the new locations

Good point. See #83.

klemens commented:
Merged! Should be up *soon*.
